content/posts/numpy/numpy-rng/index.md

```diff
@@ -0,0 +1,144 @@
+
+```python
+import numpy as np
+
+rng = np.random.default_rng()
+rng.random()  # generate a floating point number between 0 and 1
+```
+
+If you want to use a seed for reproducibility, [the NumPy documentation recommends](https://numpy.org/doc/stable/reference/random/index.html#quick-start) using a very large and unique number to have a different seed than anyone else. Another reason is that using a small number for the seed could lead to a bad first internal state of the RNG leading to a sequence of numbers that would not look like a random sequence, although `default_rng` now takes care of creating a good internal state from a small number. To obtain a good and unique seed you can rely on the [secrets module](https://docs.python.org/3/library/secrets.html). The following code will return a 128 bit number that you can then use as a seed:
+
+```python
+import secrets
+
+secrets.randbits(128)
+```
+
+When running this code I get `65647437836358831880808032086803839626` for the number to use as my seed. This number is randomly generated so you need to copy paste the value that is returned by `secrets.randbits(128)` otherwise you will have a different seed each time you run your code and thus break reproducibility:
+
+```python
+import numpy as np
+
+seed = 65647437836358831880808032086803839626
+rng = np.random.default_rng(seed)
+rng.random()
+```
+
+The reason for seeding your RNG only once (and passing that RNG around) is that with a good RNG such as the one returned by `default_rng` you will be ensured good randomness and independence of the generated numbers. However, if not done properly, using several RNGs (each one created with its own seed) might lead to streams of random numbers that are less independent than the ones created from the same seed. That being said, [as explained by Robert Kern](https://github.com/numpy/numpy/issues/15322#issuecomment-573890207), with the RNGs and seeding strategies introduced in NumPy 1.17, it is considered fairly safe to create RNGs using system entropy, i.e. using `default_rng(None)` multiple times. However as explained later be careful when running jobs in parallel and relying on `default_rng(None)`. Another reason for seeding your RNG only once is that obtaining a good seed can be time consuming. Once you have a good seed to instantiate your generator, you might as well use it.
+
+## Passing a NumPy RNG around
+
+As you write functions that you will use on their own as well as in a more complex script it is convenient to be able to pass a seed or your already created RNG. The function `default_rng` allows you to do this very easily. As written above, this function can be used to create a new RNG from your chosen seed, if you pass a seed to it, or from system entropy when passing `None` but you can also pass an already created RNG. In this case the returned RNG is the one that you passed.
+
+```python
+import numpy as np
+
+
+def stochastic_function(rng, high=10):
```

Comment:
Typical signature would be `stochastic_function(high=10, rng=None)` or similar?

Reply:
Yes true I'll fix this